crates/sui-indexer-alt-framework/src/lib.rs

```diff
@@ -363,21 +449,90 @@ impl<T: TransactionalStore> Indexer<T> {
 
 #[cfg(test)]
 mod tests {
-    use super::*;
+    use std::sync::Arc;
+    use std::sync::atomic::{AtomicU64, Ordering};
+
+    use async_trait::async_trait;
+    use sui_synthetic_ingestion::synthetic_ingestion;
+    use tokio_util::sync::CancellationToken;
+
     use crate::FieldCount;
     use crate::ingestion::ingestion_client::IngestionClientArgs;
     use crate::mocks::store::MockStore;
     use crate::pipeline::{Processor, concurrent::ConcurrentConfig};
     use crate::store::CommitterWatermark;
-    use async_trait::async_trait;
-    use std::sync::Arc;
-    use sui_synthetic_ingestion::synthetic_ingestion;
-    use tokio_util::sync::CancellationToken;
+
+    use super::*;
 
     #[allow(dead_code)]
     #[derive(Clone, FieldCount)]
     struct MockValue(u64);
 
+    /// A handler that can be controlled externally to block checkpoint processing.
+    struct ControllableHandler {
+        /// Process up to this checkpoint number (inclusive).
+        pause_at_cp: Arc<AtomicU64>,
```

Comment:
Here I think you want a `watch` channel,


```suggestion
        pause_at_cp: watch::Receiver<u64>
```

You initialize it with a receiver, where you can control the sender centrally and the process function includes a `wait_for` call, waiting until the watch channel is sent a value that is at or above the checkpoint being processed.

This avoids you having to hand-roll the "wait for a condition to hold" logic, using manual polling.